Library/Homebrew/api.rb

```diff
@@ -92,21 +92,62 @@ def self.fetch_json_api_file(endpoint, target:)
 
         retry
       end
+
+      if jws
+        success, data = verify_and_parse_jws(json_data)
+        unless success
+          target.unlink
+          odie "Failed to verify integrity of #{url} (#{data})."
+        end
+        [data, !skip_download]
+      else
+        [json_data, !skip_download]
+      end
+    end
+
+    sig { params(json_data: Hash).returns([T::Boolean, T.any(String, Array, Hash)]) }
+    def self.verify_and_parse_jws(json_data)
+      signatures = json_data["signatures"]
+      homebrew_signature = signatures&.find { |sig| sig.dig("header", "kid") == "homebrew-1" }
+      return false, "key not found" if homebrew_signature.nil?
+
+      header = JSON.parse(Base64.urlsafe_decode64(homebrew_signature["protected"]))
+      if header["alg"] != "PS512" || header["b64"] != false # NOTE: nil has a meaning of true
+        return false, "invalid algorithm"
+      end
+
+      require "openssl"
+
+      pubkey = OpenSSL::PKey::RSA.new(File.read("#{__dir__}/api/homebrew-1.pem"))
+      signing_input = "#{homebrew_signature["protected"]}.#{json_data["payload"]}"
+      unless pubkey.verify_pss("SHA512",
+                               Base64.urlsafe_decode64(homebrew_signature["signature"]),
+                               signing_input,
+                               salt_length: :digest,
+                               mgf1_hash:   "SHA512")
+        return false, "signature mismatch"
+      end
+
+      [true, JSON.parse(json_data["payload"])]
     end
+    private_class_method :verify_and_parse_jws
```

Comment:
Might be nicer to put this at the bottom of the file; that's where private methods typically reside.